### PR TITLE
fix audio codec issue

### PR DESCRIPTION
**Pull request by niyatim23**
*Issue #, if available:*
https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/issues/229
*Description of changes:*
Fix the issue from SDK side by setting the correct audio codec and content type
Changed the _kvsAacAudioVideoStreamingSample_ to work with both _AAC_ and _PCM Alaw._ 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by niyatim23** [comment]
This is to resolve the G.711 issue raised by @awsa2ron [here](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/issues/230)

File: src/source/StreamInfoProvider.h [hunk 1]
```diff
@@ -64,21 +68,24 @@ STATUS setStreamInfoDefaults(STREAMING_TYPE, UINT64, UINT64, UINT32, PStreamInfo
...
  * @param - PCHAR - stream name
  * @param - UINT64 - retention
  * @param - UINT64 - bufferDuration
+ * @param - VIDEO_CODEC_ID Codec ID of the video(H.264/H.265)
+ * @param - AUDIO_CODEC_ID Codec ID of the audio(AAC/PCM_ALAW/PCM_MULAW)
  * @param - PStreamInfo
  * @return - STATUS code of the execution
  */
-STATUS createAudioVideoStreamInfo(STREAMING_TYPE, PCHAR, UINT64, UINT64, PStreamInfo*);
+STATUS createAudioVideoStreamInfo(STREAMING_TYPE, PCHAR, UINT64, UINT64, VIDEO_CODEC_ID, AUDIO_CODEC_ID, PStreamInfo*);
```
Comment (disa6302) [inline src/source/StreamInfoProvider.h:88]: For the sake of uniformity, let us move the enums to the start of the signature. Something like this: 
`STATUS createAudioVideoStreamInfo(STREAMING_TYPE, VIDEO_CODEC_ID, AUDIO_CODEC_ID, PCHAR, UINT64, UINT64, PStreamInfo*);`

File: src/source/StreamInfoProvider.h [hunk 2]
```diff
@@ -64,21 +68,24 @@ STATUS setStreamInfoDefaults(STREAMING_TYPE, UINT64, UINT64, UINT32, PStreamInfo
  * @param - PCHAR  - stream name
  * @param - UINT64 - retention
  * @param - UINT64 - bufferDuration
+ * @param - VIDEO_CODEC_ID Codec ID of the video(H.264/H.265)
  * @param - PStreamInfo
  * @return - STATUS code of the execution
  */
-STATUS createVideoStreamInfo(STREAMING_TYPE, PCHAR, UINT64, UINT64, PStreamInfo*);
+STATUS createVideoStreamInfo(STREAMING_TYPE, PCHAR, UINT64, UINT64, VIDEO_CODEC_ID, PStreamInfo*);
```
Comment (disa6302) [inline src/source/StreamInfoProvider.h:75]: For the sake of uniformity, let us move the enums to the start of the signature. Something like this: 
`STATUS createAudioVideoStreamInfo(STREAMING_TYPE, VIDEO_CODEC_ID, PCHAR, UINT64, UINT64, PStreamInfo*);`

File: src/source/StreamInfoProvider.c [hunk 3]
```diff
@@ -49,31 +70,59 @@ STATUS freeStreamInfoProvider(PStreamInfo* ppStreamInfo)
...
+// Creates track info for audio with given codec and sets the content type as per the codec
+STATUS createAudioTrackInfo(AUDIO_CODEC_ID audioCodecId, PCHAR contentType, PTrackInfo pTrackInfo)
 {
     STATUS retStatus = STATUS_SUCCESS;
 
     pTrackInfo->trackId = DEFAULT_AUDIO_TRACK_ID;
     pTrackInfo->codecPrivateData = NULL;
     pTrackInfo->codecPrivateDataSize = 0;
-    STRCPY(pTrackInfo->codecId, MKV_AAC_CODEC_ID);
+
```
Comment (disa6302) [inline src/source/StreamInfoProvider.c:108]: Let us add a NULL check for `contentType` and `pTrackInfo`. Something like:


`CHK(contentType != NULL, STATUS_NULL_ARG)`. Similarly for `pTrackInfo`

File: src/source/StreamInfoProvider.c [hunk 4]
```diff
@@ -49,31 +70,59 @@ STATUS freeStreamInfoProvider(PStreamInfo* ppStreamInfo)
     LEAVES();
     return retStatus;
 }
-
-// Creates H264 track info for video
-STATUS createH264VideoTrackInfo(PTrackInfo pTrackInfo)
+// Creates track info for video with given codec and sets the content type as per the codec
+STATUS createVideoTrackInfo(VIDEO_CODEC_ID videoCodecId, PCHAR contentType, PTrackInfo pTrackInfo)
```
Comment (disa6302) [inline src/source/StreamInfoProvider.c:74]: `NULL` check the args here as well.

File: src/source/StreamInfoProvider.c (hunk 3 above)
Comment (niyatim23) [reply src/source/StreamInfoProvider.c:108]: Added the `CHK(pTrackInfo != NULL && pStreamInfo->streamCaps.contentType != NULL, STATUS_NULL_ARG);` to `createAudioVideoStreamInfo` and `createVideoStreamInfo` which call this function

File: src/source/StreamInfoProvider.c (hunk 4 above)
Comment (niyatim23) [reply src/source/StreamInfoProvider.c:74]: Added the `CHK(pTrackInfo != NULL && pStreamInfo->streamCaps.contentType != NULL, STATUS_NULL_ARG);` to `createAudioVideoStreamInfo` and `createVideoStreamInfo` which call this function

File: src/source/StreamInfoProvider.h (hunk 2 above)
Comment (niyatim23) [reply src/source/StreamInfoProvider.h]: Updated as mentioned in the new commit

File: src/source/StreamInfoProvider.h (hunk 1 above)
Comment (niyatim23) [reply src/source/StreamInfoProvider.h]: Updated as mentioned in the new commit

File: samples/KvsAudioVideoStreamingSample.c [hunk 5]
```diff
@@ -169,10 +174,16 @@ INT32 main(INT32 argc, CHAR* argv[])
     UINT32 i;
     CHAR filePath[MAX_PATH_LEN + 1];
     PTrackInfo pAudioTrack = NULL;
-    BYTE audioCpd[KVS_AAC_CPD_SIZE_BYTE];
+    CHAR audioCodec[AUDIO_CODEC_NAME_MAX_LENGTH] = AUDIO_CODEC_NAME_AAC; //aac audio by default
```
[the remark this answers is not in the document]
Comment (niyatim23) [reply samples/KvsAudioVideoStreamingSample.c]: Updated as mentioned in the new commit

File: README.md [hunk 6]
```diff
@@ -70,10 +70,13 @@ To run the samples:
 export AWS_SECRET_ACCESS_KEY=<YourAWSSecretAccessKey>
 export AWS_ACCESS_KEY_ID=<YourAWSAccessKey>
 ```
-For audio+video, run `./kvsAacAudioVideoStreamingSample <channel-name`
+For audio+video, run `./kvsAudioVideoStreamingSample <channel-name`
```
[the remark this answers is not in the document]
Comment (niyatim23) [reply README.md]: Updated and added more information in the new commit

File: src/source/StreamInfoProvider.c [hunk 7]
```diff
@@ -49,31 +70,59 @@ STATUS freeStreamInfoProvider(PStreamInfo* ppStreamInfo)
...
     STATUS retStatus = STATUS_SUCCESS;
 
     pTrackInfo->trackId = DEFAULT_VIDEO_TRACK_ID;
     pTrackInfo->codecPrivateData = NULL;
     pTrackInfo->codecPrivateDataSize = 0;
-    STRCPY(pTrackInfo->codecId, MKV_H264_AVC_CODEC_ID);
+    switch (videoCodecId) {
+        case VIDEO_CODEC_ID_H264:
+            STRCPY(pTrackInfo->codecId, MKV_H264_AVC_CODEC_ID);
+            STRCAT(contentType, MKV_H264_CONTENT_TYPE);
```
Comment (hassanctech) [inline src/source/StreamInfoProvider.c:84]: I see this works because we calloc the streaminfo stuct so the first byte is all 0's, i.e. null char.

**Review by hassanctech — APPROVED** [review]
Looks great @niyatim23!  Thanks for addressing all the comments from @disa6302.

File: README.md [hunk 8]
```diff
@@ -70,10 +70,19 @@ To run the samples:
...
+* the `streaming-duration-in-seconds` is 20 seconds
+* `sample-location` is `../samples`
+* `audio-codec` is `aac`
+
+If you want to use the sample for `PCM_ALAW/G.711` frames, run 
+`./kvsAudioVideoStreamingSample <channel-name> <streaming_duration> <sample_location> alaw`
+
+For video only, run `./kvsVideoOnlyStreamingSample <channel-name>`
+
+This will stream the video/audio files from the `samples/h264SampleFrames` and `samples/aacSampleFrames` respectively by default.
```
Comment (disa6302) [inline README.md:85]: Missed this yest, for video only, `samples/aacSampleFrames` will not be used.

**Review by disa6302 — APPROVED** [review]
Great job @niyatim23 ! Just a minor documentation comment. Once that is fixed, the PR is good to go.

File: src/source/StreamInfoProvider.c (hunk 4 above)
Comment (jdelapla) [reply src/source/StreamInfoProvider.c:74]: Yes but this function is in the header file, and can be called by other functions we may develop in the future. It should have checks as well.
Add NULL checks here and in createAudioTrackInfo.

**Review by jdelapla — CHANGES_REQUESTED** [review]

**Comment by jdelapla** [comment]
Very well done overall. Please add the indent and NULL checks where specified.

File: samples/KvsAudioVideoStreamingSample.c [hunk 9]
```diff
@@ -235,19 +249,36 @@ INT32 main(INT32 argc, CHAR* argv[])
...
+        // set up audio cpd.
+        pAudioTrack = pStreamInfo->streamCaps.trackInfoList[0].trackId == DEFAULT_AUDIO_TRACK_ID ? &pStreamInfo->streamCaps.trackInfoList[0]
                                                                                              : &pStreamInfo->streamCaps.trackInfoList[1];
-    // generate audio cpd
-    pAudioTrack->codecPrivateData = audioCpd;
-    pAudioTrack->codecPrivateDataSize = KVS_AAC_CPD_SIZE_BYTE;
-    CHK_STATUS(mkvgenGenerateAacCpd(AAC_LC, AUDIO_TRACK_SAMPLING_RATE, AUDIO_TRACK_CHANNEL_CONFIG, pAudioTrack->codecPrivateData,
+        pAudioTrack->codecPrivateData = alawAudioCpd;
+        pAudioTrack->codecPrivateDataSize = KVS_PCM_CPD_SIZE_BYTE;
+    CHK_STATUS(mkvgenGeneratePcmCpd(KVS_PCM_FORMAT_CODE_ALAW, ALAW_AUDIO_TRACK_SAMPLING_RATE, ALAW_AUDIO_TRACK_CHANNEL_CONFIG, pAudioTrack->codecPrivateData,
```
[the remark this answers is not in the document]
Comment (niyatim23) [reply samples/KvsAudioVideoStreamingSample.c]: Added the indentation

File: README.md (hunk 8 above)
Comment (niyatim23) [reply README.md]: removed `samples/aacSampleFrames` from this section and added it to the audio+video section

File: src/source/StreamInfoProvider.c (hunk 4 above)
Comment (niyatim23) [reply src/source/StreamInfoProvider.c:74]: Right, I added the NULL checks in `createAudioTrackInfo` and `createVideoTrackInfo` as well

**Review by jdelapla — APPROVED** [review]

**Pull request merged**